File: Eulers-Method/Eulers_Method.py
# solution to the differential equation y' = y - t^2 + 1, y(0) = 0.5, with 0<=t<=2
# exact solution for this differential equaiton is y = (t + 1)^2 - 0.5 * exp(t)
# for more accurate results decrease the step size

from typing import Any
import numpy as np
from matplotlib import pyplot as plt

# defining the IVP
def dydt(t, y):
    y1: Any = y - t ** 2 + 1
    return y1

# exact solution
def exact_sol(t):
    y = (t + 1) ** 2 - 0.5 * np.exp(t)
    return y

# initial values
t0 = 0.0
y0 = 0.5
t_final = 2
h = 0.1 # step size
n = int((t_final - t0) / h)
t_val = np.zeros(n + 1)
y_val = np.zeros(n + 1)
t_val[0] = t0
y_val[0] = y0

# applying the Euler's Method
for i in range(1, n + 1):
    t_val[i] = round(t_val[i - 1] + h, 3)
    y_val[i] = round(y_val[i - 1] + h * dydt(t_val[i - 1], y_val[i - 1]), 7)

# printing the solution
for i in range(0, n + 1):
    exact = round(exact_sol(t_val[i]), 7)
    print("t: ", t_val[i], "\ty: ", y_val[i], "\t\texact solution: ", exact)

# plotting the solution
plt.figure('Solution')
plt.plot(t_val, y_val, 'r-o')
plt.grid(True)
plt.xlabel('t')
plt.ylabel('y')
plt.legend('N')
plt.title('Numeric Solution to a IVP')
plt.show()

File: Eulers-Method/Eulers-Method.m
clear all
% using clear all at the beginning to avoid conflict among variables

% initializing symbols
syms exact_sol t;

% defining the ODE function y' = y - t^2 + 1
dydt = @(t, y) y - t^2 + 1;

% defining the initial conditions
t0 = 0;
y0 = 0.5;

% defining the final value of x
t_final = 2;

% step size
step = 0.1;

% initializing the arrays to store the variables
t_val = t0 : step : t_final;
y_val = zeros(size(t_val));
y_val(1) = y0;

% applying Euler's method
for i = 1 : length(t_val) - 1
    y_val(i + 1) = y_val(i) + step * dydt(t_val(i), y_val(i));
end

% exact solution
exact_sol = (t + 1)^2 - 0.5 * exp(t);

% displaying the solutions
for i = 1 : length(t_val)
    exact = subs(exact_sol, t_val(i));
    fprintf("t: %0.2f \ty: %0.7f\t exact: %0.7f\n", t_val(i), y_val(i), exact);
end

% plotting both the solutions
% numerical plot
% plotting both the solutions
plot(t_val, y_val, 'b-o');
hold on;
% symbolic plot
fplot(exact_sol, [0 2], 'r');
title('Numerical solution using Euler''s method');
xlabel('t');
ylabel('y');
legend('numeric', 'analytic');
